services/s3/s3.go

```diff
@@ -173,6 +174,130 @@ func (s *S3) HeadObject(input GetObjectInput) (*GetObjectOutput, *awserrors.Erro
 	return s.getObject(input, false)
 }
 
+type ByteRange struct {
+	startByte int64
+	endByte   int64
+}
+
+func parseRangeHeader(rangeHeader string, o *Object) ([]ByteRange, *awserrors.Error) {
+	bytesPrefix := "bytes="
+	if !strings.HasPrefix(rangeHeader, bytesPrefix) {
+		awserrors.XXX_TODO("unknown range header unit")
+	}
+
+	ranges := strings.Split(rangeHeader[len(bytesPrefix):], ",")
+	var result []ByteRange
+
+	for _, encodedRange := range ranges {
+		trimmed := strings.TrimSpace(encodedRange)
+		splitRange := strings.Split(trimmed, "-")
+		if len(splitRange) != 2 {
+			return nil, awserrors.XXX_TODO("invalid range too many -")
+		}
+
+		// indexed from end of resource
+		if splitRange[0] == "" {
+			suffix, err := strconv.ParseInt(splitRange[1], 10, 64)
+			if err != nil {
+				return nil, awserrors.XXX_TODO(err.Error())
+			}
+			result = append(result, ByteRange{
+				startByte: o.ContentLength - suffix,
+				endByte:   o.ContentLength,
```

Comment:
might make a bit more sense to convert to exclusive endpoint here as you parse, otherwise you account for it in multiple places later